runtime/fundamentals/testing.md

```diff
@@ -0,0 +1,757 @@
+find and execute all tests in the current directory (recursively) that match the
+glob `{*_,*.,}test.{ts, tsx, mts, js, mjs, jsx}`.
+
+```sh
+# Run all tests in the current directory and all sub-directories
+deno test
+
+# Run all tests in the util directory
+deno test util/
+
+# Run just my_test.ts
+deno test my_test.ts
+
+# Run test modules in parallel
+deno test --parallel
+
+# Pass additional arguments to the test file
+deno test my_test.ts -- -e --foo --bar
+```
+
+## Test Steps
+
+Deno also supports test steps, which allow you to break down tests into smaller,
+manageable parts. This is useful for setup and teardown operations within a
+test:
+
+```ts
+Deno.test("database operations", async (t) => {
+  await t.step("insert user", async () => {
+    // Insert user logic
+  });
+  await t.step("insert book", async () => {
+    // Insert book logic
+  });
+});
+```
+
+## Command line filtering
+
+Deno allows you to run specific tests or groups of tests using the `--filter`
+option on the command line. This option accepts either a string or a pattern to
+match test names.
+
+Consider the following tests:
+
+```ts
+Deno.test({ name: "my-test", fn: myTest });
+Deno.test({ name: "test-1", fn: test1 });
+Deno.test({ name: "test-2", fn: test2 });
+```
+
+### Filtering by string
+
+To run all tests that contain the word "test" in their names, use:
+
+```sh
+deno test --filter "test" tests/
+```
+
+This command will execute `my-test`, `test-1`, and `test-2` because they all
+include "test" in their names.
+
+### Filtering by Pattern
+
+To run tests that match a specific pattern, use:
+
+```sh
+deno test --filter "/test-*\d/" tests/
+```
+
+This command will run `test-1` and `test-2` because they match the pattern
+`test-*` followed by a digit.
+
+To indicate that you are using a pattern (regular expression), wrap your filter
+value with forward slashes `/`, much like JavaScript’s syntax for regular
+expressions.
+
+### Including and excluding paths in the configuration file
+
+You can also filter tests by specifying paths to include or exclude in the Deno
+configuration file.
+
+For example, if you want to only test `src/fetch_test.ts` and
+`src/signal_test.ts` and exclude everything in `out/`:
+
+```json
+{
+  "test": {
+    "include": [
+      "src/fetch_test.ts",
+      "src/signal_test.ts"
+    ]
+  }
+}
+```
+
+Or more likely:
+
+```json
+{
+  "test": {
+    "exclude": ["out/"]
+  }
+}
+```
+
+Then running `deno test` in the same directory tree as the configuration file
+will take these options into account.
+
+## Test definition filtering
+
+Deno provides two options for filtering tests within the test definitions
+themselves: ignoring tests and focusing on specific tests.
+
+### Filtering Out (Ignoring Tests)
+
+You can ignore certain tests based on specific conditions using the `ignore`
+boolean in the test definition. If `ignore` is set to `true`, the test will be
+skipped. This is useful, for example, if you only want a test to run on a
```

Comment:
We should add a sentence that it's still shown as being skipped in the test runner output. Maybe even include the output of the terminal directly here as an example, or later after the next code snippet.